ArduCopter/autoyaw.cpp

```diff
@@ -0,0 +1,213 @@
+#if MOUNT == ENABLED
+        // switch off the camera tracking if enabled
+        if (copter.camera_mount.get_mode() == MAV_MOUNT_MODE_GPS_POINT) {
+            copter.camera_mount.set_mode_to_default();
+        }
+#endif  // MOUNT == ENABLED
+    } else {
+#if MOUNT == ENABLED
+        // check if mount type requires us to rotate the quad
+        if (!copter.camera_mount.has_pan_control()) {
+            roi = copter.pv_location_to_vector(roi_location);
+            set_mode(AUTO_YAW_ROI);
+        }
+        // send the command to the camera mount
+        copter.camera_mount.set_roi_target(roi_location);
+
+        // TO-DO: expand handling of the do_nav_roi to support all modes of the MAVLink.  Currently we only handle mode 4 (see below)
+        //      0: do nothing
+        //      1: point at next waypoint
+        //      2: point at a waypoint taken from WP# parameter (2nd parameter?)
+        //      3: point at a location given by alt, lon, lat parameters
+        //      4: point at a target given a target id (can't be implemented)
+#else
+        // if we have no camera mount aim the quad at the location
+        roi = pv_location_to_vector(roi_location);
+        set_mode(AUTO_YAW_ROI);
+#endif  // MOUNT == ENABLED
+    }
+}
+
+// set auto yaw rate in centi-degrees per second
+void Copter::Mode::_AutoYaw::set_rate(float turn_rate_cds)
+{
+    set_mode(AUTO_YAW_RATE);
+    _rate_cds = turn_rate_cds;
+}
+
+// yaw - returns target heading depending upon auto_yaw_mode
+float Copter::Mode::_AutoYaw::yaw()
+{
+    switch (_mode) {
+
+    case AUTO_YAW_ROI:
+        // point towards a location held in roi
+        return roi_yaw();
+
+    case AUTO_YAW_FIXED:
+        // keep heading pointing in the direction held in fixed_yaw
+        // with no pilot input allowed
+        return _fixed_yaw;
+
+    case AUTO_YAW_LOOK_AHEAD:
+        // Commanded Yaw to automatically look ahead.
+        return look_ahead_yaw();
+
+    case AUTO_YAW_RESETTOARMEDYAW:
+        // changes yaw to be same as when quad was armed
+        return copter.initial_armed_bearing;
+
+    case AUTO_YAW_LOOK_AT_NEXT_WP:
+    default:
+        // point towards next waypoint.
+        // we don't use wp_bearing because we don't want the copter to turn too much during flight
+        return copter.wp_nav->get_yaw();
+    }
+}
+
+// returns yaw rate normally set by SET_POSITION_TARGET mavlink
+// messages (positive is clockwise, negative is counter clockwise)
+float Copter::Mode::_AutoYaw::rate_cds() const
+{
+    if (_mode == AUTO_YAW_RATE) {
+        return _rate_cds;
+    }
+
+    // return zero turn rate (this should never happen)
+    return 0.0f;
+}
+
+/*************************************************************
+ * yaw controllers
+ *************************************************************/
+
+// roi_yaw - returns heading towards location held in roi
+float Copter::Mode::_AutoYaw::roi_yaw()
+{
+    roi_yaw_counter++;
+    if (roi_yaw_counter >= 4) {
```

Comment:
The idea here is to reduce the rate to 100 Hz, but it is hardcoded and will give different results if the loop speed changes.
So how about calculating the decimation value once on initialization and then keep the result in a private variable (_roi_yaw_update_decimation = LOOP_RATE_PARAMETER / 100; )

Reply:
Ah, and add a comment here explaining that this is to reduce the update rate to 100Hz